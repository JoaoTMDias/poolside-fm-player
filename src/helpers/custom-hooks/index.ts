export * from "./use-event";
export * from "./use-lifecycle-hooks";
export * from "./use-debounced-effect";
