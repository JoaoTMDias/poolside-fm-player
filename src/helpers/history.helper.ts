import { createBrowserHistory } from "history";

export const history = createBrowserHistory({
	basename: process.env.REACT_APP_BASENAME,
});

export default history;
