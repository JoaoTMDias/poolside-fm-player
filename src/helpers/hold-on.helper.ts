/**
 * Returns a promise that resolves after how many milliseconds you pass it.
 *
 * @author Jo√£o Dias
 * @param {number} [amount=0]
 * @returns
 */
export function holdOn(amount = 0): Promise<NodeJS.Timeout> {
	return new Promise((resolve) => setTimeout(resolve, amount));
}

export default holdOn;
