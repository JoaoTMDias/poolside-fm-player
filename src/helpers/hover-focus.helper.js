import { css } from "styled-components";

export const hoverFocus = styles =>
	css`
		&:hover:enabled,
		&:focus:enabled {
			${css(...styles)}
		}
	`;

export default hoverFocus;
