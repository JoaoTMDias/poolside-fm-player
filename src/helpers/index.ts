import hasPressedSpaceOrEnter from "./has-pressed-space-or-enter";

export * from "./logger.helper";
export * from "./key-codes";
export * from "./with-memo.helper";
export * from "./hold-on.helper";
export * from "./hover-focus.helper";
export * from "./history.helper";

export { hasPressedSpaceOrEnter };
