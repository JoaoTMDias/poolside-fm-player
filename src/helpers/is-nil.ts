export const isNil = (val: any) => val == null;
