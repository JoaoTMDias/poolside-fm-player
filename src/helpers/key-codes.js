export const KEY_CODES = {
	BACKSPACE: 8,
	TAB: 9,
	ENTER: 13,
	OK: 13,
	SHIFT: 16,
	ESC: 27,
	SPACE: 32,
	END: 35,
	HOME: 36,
	ARROW_LEFT: 37,
	ARROW_UP: 38,
	ARROW_RIGHT: 39,
	ARROW_DOWN: 40,
	ONE: 49,
	TWO: 50,
	THREE: 51,
	FOUR: 52,
	FIVE: 53,
	SIX: 54,
	SEVEN: 55,
	RED: 403,
	GREEN: 404,
	YELLOW: 405,
	BLUE: 406,
	BACKWARD: 412,
	FORWARD: 417,
	BACK: 461,
	PLAY_PAUSE: 463,
};

export default KEY_CODES;
