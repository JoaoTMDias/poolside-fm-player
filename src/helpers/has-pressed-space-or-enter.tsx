import { KEY_CODES } from "./key-codes";

/**
 * Checks if the user has pressed space or enter
 *
 * @param {number} keyCode
 * @returns {boolean}
 */
function hasPressedSpaceOrEnter(keyCode: number) {
	if (keyCode === KEY_CODES.ENTER || keyCode === KEY_CODES.SPACE) {
		return true;
	}

	return false;
}

export default hasPressedSpaceOrEnter;
