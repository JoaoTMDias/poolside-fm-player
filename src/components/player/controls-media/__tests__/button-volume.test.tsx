import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { KEY_CODES } from "helpers";
import ButtonVolume from "../button-volume";

afterEach(cleanup);

describe("<ButtonVolume />", () => {
	describe("render", () => {
		it("should render without errors", () => {
			const wrapper = render(<ButtonVolume />);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe("data-volume", () => {
		it("should have been updated on click on range input", async () => {
			const ButtonVolumeWithPortal = () => (
				<div id="wrapper">
					<div id="app">
						<ButtonVolume />
					</div>
					<div id="app-portal" />
				</div>
			);

			const { getByTestId } = render(<ButtonVolumeWithPortal />);
			const button = await getByTestId("component-controls-media-button-volume");

			fireEvent.click(button);

			const input = await getByTestId("component-button-volume-range-input");

			fireEvent.change(input, {
				target: {
					value: 0.5,
				},
			});

			expect(button.getAttribute("data-volume")).toBe("0.5");
		});
	});

	describe("onChangeVolume", () => {
		it("should have been called on change on range input", async () => {
			const onChangeVolumeMock = jest.fn();
			const EXPECTED_VOLUME = "0.5";

			const ButtonVolumeWithPortal = () => (
				<div id="wrapper">
					<div id="app">
						<ButtonVolume
							onChangeVolume={onChangeVolumeMock}
						/>
					</div>
					<div id="app-portal" />
				</div>
			);

			const { getByTestId } = render(<ButtonVolumeWithPortal />);
			const button = await getByTestId("component-controls-media-button-volume");

			fireEvent.click(button);

			const input = await getByTestId("component-button-volume-range-input");

			fireEvent.change(input, {
				target: {
					value: EXPECTED_VOLUME,
				},
			});

			expect(onChangeVolumeMock).toHaveBeenCalledWith(EXPECTED_VOLUME);
		});
	});

	describe("onClick", () => {
		it("should have been called on change on range input", async () => {
			const onClickMock = jest.fn();

			const ButtonVolumeWithPortal = () => (
				<div id="wrapper">
					<div id="app">
						<ButtonVolume
							onClick={onClickMock}
						/>
					</div>
					<div id="app-portal" />
				</div>
			);

			const { getByTestId } = render(<ButtonVolumeWithPortal />);
			const button = await getByTestId("component-controls-media-button-volume");

			fireEvent.click(button);

			const closeButton = await getByTestId("component-portal-click-trap");

			fireEvent.click(closeButton);

			expect(onClickMock).toHaveBeenCalled();
		});

		describe("should have been called on change on key up", () => {
			it("on ENTER input", async () => {
				const onClickMock = jest.fn();
	
				const ButtonVolumeWithPortal = () => (
					<div id="wrapper">
						<div id="app">
							<ButtonVolume
								onClick={onClickMock}
							/>
						</div>
						<div id="app-portal" />
					</div>
				);
	
				const { getByTestId } = render(<ButtonVolumeWithPortal />);
				const button = await getByTestId("component-controls-media-button-volume");
	
				fireEvent.keyUp(button, {
					keyCode: KEY_CODES.ENTER,
				});
	
				expect(onClickMock).toHaveBeenCalled();
			});

			it("on SPACE input", async () => {
				const onClickMock = jest.fn();

				const ButtonVolumeWithPortal = () => (
					<div id="wrapper">
						<div id="app">
							<ButtonVolume
								onClick={onClickMock}
							/>
						</div>
						<div id="app-portal" />
					</div>
				);

				const { getByTestId } = render(<ButtonVolumeWithPortal />);
				const button = await getByTestId("component-controls-media-button-volume");

				fireEvent.keyUp(button, {
					keyCode: KEY_CODES.SPACE,
				});

				expect(onClickMock).toHaveBeenCalled();
			});
		});

		it("should not have been called on change on key up with any other key", async () => {
			const onClickMock = jest.fn();
	
			const ButtonVolumeWithPortal = () => (
				<div id="wrapper">
					<div id="app">
						<ButtonVolume
							onClick={onClickMock}
						/>
					</div>
					<div id="app-portal" />
				</div>
			);
	
			const { getByTestId } = render(<ButtonVolumeWithPortal />);
			const button = await getByTestId("component-controls-media-button-volume");
	
			fireEvent.keyUp(button, {
				keyCode: KEY_CODES.ARROW_DOWN,
			});

			fireEvent.keyUp(button, {
				keyCode: KEY_CODES.BACK,
			});

			fireEvent.keyUp(button, {
				keyCode: KEY_CODES.PLAY_PAUSE,
			});
	
			expect(onClickMock).not.toHaveBeenCalled();
		});
	});
});
