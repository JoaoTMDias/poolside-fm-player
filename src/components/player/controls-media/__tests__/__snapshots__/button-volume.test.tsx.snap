// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonVolume /> render should render without errors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="Click to control the volume of the song"
        class="controls-media__button button button--large"
        data-testid="component-controls-media-button-volume"
        data-volume="1"
        id="controls-button-volume"
        title="Control Volume"
        type="button"
      >
        <svg
          class="icon"
          fill="none"
          height="9"
          viewBox="0 0 9 9"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.652.587V0H5.478v.587h-.587v.587h-.587v.587h-.587v.587h-1.37v.587h-.586v3.13h.587v.587h1.37v.587h.586v.587h.587v.587h.587V9h1.174v-.587h.587V.587h-.587z"
            fill="var(--color-icon, #000)"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="Click to control the volume of the song"
      class="controls-media__button button button--large"
      data-testid="component-controls-media-button-volume"
      data-volume="1"
      id="controls-button-volume"
      title="Control Volume"
      type="button"
    >
      <svg
        class="icon"
        fill="none"
        height="9"
        viewBox="0 0 9 9"
        width="9"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.652.587V0H5.478v.587h-.587v.587h-.587v.587h-.587v.587h-1.37v.587h-.586v3.13h.587v.587h1.37v.587h.586v.587h.587v.587h.587V9h1.174v-.587h.587V.587h-.587z"
          fill="var(--color-icon, #000)"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
